### PR TITLE
Remove Regex dependency

### PR DESCRIPTION
**Pull request by RossSmyth**
Regex being the slowest dependency for my entire project just to do a couple matches has been really bugging me. So I just recreated it.

The main draw back is the Unicode table. I just tried to recreate the semantics of the regex, and this is what it specified (any word character).

This is even lighter weight than regex-lite so I would say this succeeds #168, plus there is no MSRV junk to worry about (wrt dependencies). 

I don't personally believe there is any maintenance burden to it since the matching is pretty straightforward. 

I also added a test to test that we select the second part of serial numbers that are split by "&" as specified in one issue. That test seems to have been updated so that it no longer did that after parent IDs were added.

### CODE REVIEW

**Comment by parasyte** [comment]
💯

I have a pet peeve with static RegExs. They add a lot of overhead compared to a simple FSM.

**Comment by sirhcel** [comment]
Thank you for keeping on working on this issue @RossSmyth! The call for regex support is also imminent with #170 but in looks like we can get away there without regex or regex-lite there as well.

After a quick look at HWIDs and the MODALIAS strings from #170, I'm wondering if we actually need to deal with word boundaries according to Perl regex or if we could use a simpler definition in our context? I'm not opposed to a solution based on `PERL_WORD` but with this big step already made, isn't there a even more compact solution within reach?

**Comment by RossSmyth** [comment]
> I'm wondering if we actually need to deal with word boundaries according to Perl regex or if we could use a simpler definition in our context?

Yeah I was thinking the same, but I personally do not know what that subset is. The Windows documentation on HWID isn't really that great.

**Comment by RossSmyth** [comment]
Okie dokie. I'll modify it so.

File: src/windows/enumerate.rs [hunk 1]
```diff
@@ -504,7 +549,7 @@ fn test_parsing_usb_port_information() {
 
     assert_eq!(info.vid, 0x0403);
     assert_eq!(info.pid, 0x6001);
-    assert_eq!(info.serial_number, Some("A702TB52A".to_string()));
+    assert_eq!(info.serial_number, Some(r"A702TB52A\0000".to_string()));
```
Comment (RossSmyth) [inline src/windows/enumerate.rs:552]: With the new impl, if I interpreted what you & the pyserial regex mean correctly, it just matches to the end of the string. Meaning that this is the serial number it interprets now.

Comment (sirhcel) [reply src/windows/enumerate.rs:552]: Wow, you are quick! Yes, pySerial's regex matches until the end of the string. But they are distinguishing between the two bus types `USB` and `FTDIBUS` for the former we are sorted. But for the latter we need to stop matching at the end of the middle paths segment and ignore `\0000`.

Comment (sirhcel) [reply src/windows/enumerate.rs:552]: May be splitting the string at first at `\`, then deciding how to interpret middle path element according to the first one (`FTDIBUS` and everything else) and then handle things according to whether we are dealing with the interface of a composite device or the device itself.

**Comment by RossSmyth** [comment]
Thinking about how the serial number is matched, we currently make the serial number an optional field, even from the parent device if part of a composite device. We do not test this behavior though. So I wonder what the semantics are supposed to be. 

1. Serial numbers are always optional (this is what is currently implemented)
2. Serial numbers of child devices of composite devices are optional, but the parent will always have a serial
3. Something weirder?

**Comment by RossSmyth** [comment]
Ok, my Python skills are quite rusty but I think it's starting right in the face in your link.
https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L337-L343

I believe it is supposed to be that if the child device doesn't have a serial number, the parent must, or else pyserial will throw an exception OR stack overflow because it seems to just infinitely recurse 
https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L230-L231

**Comment by sirhcel** [comment]
> Thinking about how the serial number is matched, we currently make the serial number an optional field, even from the parent device if part of a composite device. We do not test this behavior though. So I wonder what the semantics are supposed to be.
> 
>     1. Serial numbers are always optional (this is what is currently implemented)

As far as I've overseeing things here: This is is the way. Serial numbers for USB devices are provided by the device and are optional. In terms of USB descriptors the serial number is a property of the USB device and not the interface. In case of a composite device there is one serial number per device and the interfaces are numbered.

This is also related to #178 and we are maybe using the wrong serial number for composite FTDI devices: The suffix seems to be generated by the FTDI driver and according to https://github.com/pyserial/pyserial/issues/283#issue-274725473 (COM96 and COM134), the parent device provides the right serial number.

**Comment by RossSmyth** [comment]
> the parent device provides the right serial number.

If you look at `get_parent_serial_number()` that does not seem to be the guaranteed case.
https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L162-L164

So yeah, I agree. Always optional.

**Comment by sirhcel** [comment]
> I believe it is supposed to be that if the child device doesn't have a serial number, the parent must, or else pyserial will throw an exception OR stack overflow because it seems to just infinitely recurse https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L230-L231

Recursion depth is limited [here](https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L163).


> I'll change this in a later PR though as it is a bit orthogonal.

Yes. No need to change/fix everything here. We just should end up at least en par with the beforehand state.

**Comment by sirhcel** [comment]
> If you look at `get_parent_serial_number()` that does not seem to be the guaranteed case. https://github.com/pyserial/pyserial/blob/7aeea35429d15f3eefed10bbb659674638903e3a/serial/tools/list_ports_windows.py#L162-L164

Like in: No serial if the actual device is nested too deep down in the device stack?

> So yeah, I agree. Always optional.

Great to be on the same page! 😄

**Comment by sirhcel** [comment]
I would like to keep the actual wrong serial numbers returned in #203 and not return a longer number which is also not the devices serial number.

Switching from `char::is_ascii_alphanumeric` to `char::is_alphanumeric` should bring us a bit more toward `PERL_WORD` with still just code from the standard library. Do you see an issue with that? Did you use `is_ascii_alphanumeric` for a certain reason?

Locally testing the changes shows no difference in the output for my set of test hardware. From my perspective, we are good to go. Thank you very much for your perseverance with this PR!

**Comment by RossSmyth** [comment]
I used it because it is semantically identical to the pyserial regex. But we can preserve the PERL_WORD semantics instead, that works for me.

**Comment by RossSmyth** [comment]
Oops forgot to actually add the changes 😄 

It did not handling actual Unicode in the serial number properly if we are to accept that. The previous implementation would index it by byte. For for example the test serial I added, `3854356β03432`, would be matched as `3854356β0343` since it would index it by byte and have the end chopped off. That is fixed with tests that poke it with the edge cases I found while implementing.

**Comment by sirhcel** [comment]
Thank you for finally remembering and adding the last fix!

**Comment by RossSmyth** [comment]
Thanks!
